### PR TITLE
GitHub Actions CI

### PR DESCRIPTION
**Pull request by sagudev**
Currently it builds esr102 linux x64 and makes build available for download, but I am confident I can also make Windows builds. GitHub Actions supports MacOS too and Linux arm64 should be possible using qemu.

Run time is spent ~20min on hg cloning and ~20min on building (with sccache enabled; without sccache building takes ~40min).

Test run on my fork: https://github.com/sagudev/thunderbird-patches/actions/runs/5518649173

If anybody could test build artifact on ubuntu-20.04 please do and report if it is working.

### CODE REVIEW

**Comment by sagudev** [comment]
Getting this error on 115:
```console
2023-07-11T16:47:46.6835732Z  3:11.53   running: "/opt/hostedtoolcache/sccache/0.5.4/x64/sccache" "/usr/bin/clang++" "-O1" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=x86_64-unknown-linux-gnu" "--sysroot" "/home/runner/.mozbuild/sysroot-x86_64-linux-gnu" "-std=gnu++17" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/stl_wrappers" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/system_wrappers" "-include" "/home/runner/work/thunderbird-patches/mozilla-esr115/config/gcc_hidden.h" "-U_FORTIFY_SOURCE" "-D_FORTIFY_SOURCE=2" "-fstack-protector-strong" "-DNDEBUG=1" "-DTRIMMED=1" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/toolkit/library/rust" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/toolkit/library/rust" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/include" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/include/nspr" "-I/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/include/nss" "-DMOZILLA_CLIENT" "-include" "/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/mozilla-config.h" "-fno-sized-deallocation" "-fno-aligned-new" "-fno-exceptions" "-fPIC" "-fno-rtti" "-ffunction-sections" "-fdata-sections" "-fno-exceptions" "-fno-math-errno" "-pthread" "-pipe" "-gdwarf-4" "-O2" "-fomit-frame-pointer" "-funwind-tables" "-Wno-error=tautological-type-limit-compare" "-Wno-invalid-offsetof" "-Wno-range-loop-analysis" "-Wno-error=deprecated" "-Wno-error=deprecated-anon-enum-enum-conversion" "-Wno-error=deprecated-enum-enum-conversion" "-Wno-error=deprecated-this-capture" "-Wno-inline-new-delete" "-Wno-error=deprecated-declarations" "-Wno-error=array-bounds" "-Wno-error=atomic-alignment" "-Wno-psabi" "-Wno-error=builtin-macro-redefined" "-Wno-unknown-warning-option" "-fexperimental-new-pass-manager" "-fno-strict-aliasing" "-ffp-contract=off" "-DMOZILLA_CONFIG_H" "-I" "glsl-optimizer/include" "-I" "glsl-optimizer/src/mesa" "-I" "glsl-optimizer/src/mapi" "-I" "glsl-optimizer/src/compiler" "-I" "glsl-optimizer/src/compiler/glsl" "-I" "glsl-optimizer/src/gallium/auxiliary" "-I" "glsl-optimizer/src/gallium/include" "-I" "glsl-optimizer/src" "-I" "glsl-optimizer/src/util" "-D__STDC_FORMAT_MACROS" "-D_GNU_SOURCE" "-DHAVE_ENDIAN_H" "-DHAVE_PTHREAD" "-DHAVE_TIMESPEC_GET" "-DMOZ_INCLUDE_MOZALLOC_H" "-Dmozilla_throw_gcc_h" "-o" "/home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/release/build/glslopt-3f68382a7b392874/out/glsl-optimizer/src/compiler/glsl_types.o" "-c" "glsl-optimizer/src/compiler/glsl_types.cpp"
2023-07-11T16:47:46.6841042Z  3:11.53   cargo:warning=In file included from glsl-optimizer/src/compiler/glsl_types.cpp:25:
2023-07-11T16:47:46.6842910Z  3:11.53   cargo:warning=In file included from glsl-optimizer/src/mesa/main/macros.h:35:
2023-07-11T16:47:46.6844733Z  3:11.53   cargo:warning=In file included from glsl-optimizer/src/util/u_math.h:42:
2023-07-11T16:47:46.6846564Z  3:11.53   cargo:warning=In file included from glsl-optimizer/include/c99_math.h:37:
2023-07-11T16:47:46.6848546Z  3:11.53   cargo:warning=In file included from /home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/system_wrappers/math.h:3:
2023-07-11T16:47:46.6850620Z  3:11.53   cargo:warning=/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/math.h:65:12: error: target of using declaration conflicts with declaration already in scope
2023-07-11T16:47:46.6852383Z  3:11.53   cargo:warning=using std::isinf;
2023-07-11T16:47:46.6853937Z  3:11.53   cargo:warning=           ^
2023-07-11T16:47:46.6855782Z  3:11.53   cargo:warning=/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/cmath:592:3: note: target of using declaration
2023-07-11T16:47:46.6857475Z  3:11.54   cargo:warning=  isinf(double __x)
2023-07-11T16:47:46.6859036Z  3:11.54   cargo:warning=  ^
2023-07-11T16:47:46.6861129Z  3:11.54   cargo:warning=/home/runner/.mozbuild/sysroot-x86_64-linux-gnu/usr/include/x86_64-linux-gnu/bits/mathcalls.h:201:19: note: conflicting declaration
2023-07-11T16:47:46.6862933Z  3:11.54   cargo:warning=__MATHDECL_1 (int,isinf,, (_Mdouble_ __value)) __attribute__ ((__const__));
2023-07-11T16:47:46.6864557Z  3:11.54   cargo:warning=                  ^
2023-07-11T16:47:46.6866328Z  3:11.54   cargo:warning=In file included from glsl-optimizer/src/compiler/glsl_types.cpp:25:
2023-07-11T16:47:46.6868333Z  3:11.54   cargo:warning=In file included from glsl-optimizer/src/mesa/main/macros.h:35:
2023-07-11T16:47:46.6870243Z  3:11.54   cargo:warning=In file included from glsl-optimizer/src/util/u_math.h:42:
2023-07-11T16:47:46.6872069Z  3:11.54   cargo:warning=In file included from glsl-optimizer/include/c99_math.h:37:
2023-07-11T16:47:46.6874200Z  3:11.54   cargo:warning=In file included from /home/runner/work/thunderbird-patches/mozilla-esr115/obj-x86_64-pc-linux-gnu/dist/system_wrappers/math.h:3:
2023-07-11T16:47:46.6876349Z  3:11.54   cargo:warning=/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/math.h:66:12: error: target of using declaration conflicts with declaration already in scope
2023-07-11T16:47:46.6878121Z  3:11.54   cargo:warning=using std::isnan;
2023-07-11T16:47:46.6879697Z  3:11.54   cargo:warning=           ^
2023-07-11T16:47:46.6881505Z  3:11.54   cargo:warning=/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/cmath:619:3: note: target of using declaration
2023-07-11T16:47:46.6883179Z  3:11.54   cargo:warning=  isnan(double __x)
2023-07-11T16:47:46.6884737Z  3:11.54   cargo:warning=  ^
2023-07-11T16:47:46.6886588Z  3:11.54   cargo:warning=/home/runner/.mozbuild/sysroot-x86_64-linux-gnu/usr/include/x86_64-linux-gnu/bits/mathcalls.h:234:19: note: conflicting declaration
2023-07-11T16:47:46.6888394Z  3:11.54   cargo:warning=__MATHDECL_1 (int,isnan,, (_Mdouble_ __value)) __attribute__ ((__const__));
2023-07-11T16:47:46.6890001Z  3:11.54   cargo:warning=                  ^
2023-07-11T16:47:46.6891569Z  3:11.54   cargo:warning=2 errors generated.
2023-07-11T16:47:46.6893113Z  3:11.54   exit status: 1
```

Should 115 be buildable?

**Comment by Betterbird** [comment]
We've built 115 on Windows many times and a volunteer built and earlier beta in mid-June for Debian 11 (glibc 2.31) and also aarch64. I don't know which build config he used, he does something special at least for aarch64.

We haven't built on Linux or Mac yet, this is planned for this week since TB 115.0 is apparently released today. That said, you don't miss a whole lot with 115 so far, it has some serious funcional regressions, some of which are fixed in BB.

Looking at the build error, you're getting a `math.h:66:12: error: target of using declaration conflicts with declaration already in scope`. Certainly nothing the BB patches broke.

**Comment by sagudev** [comment]
> Looking at the build error, you're getting a math.h:66:12: error: target of using declaration conflicts with declaration already in scope. Certainly nothing the BB patches broke.

Yeah, that's what I thought, I just wanted to know if it is my fault or is there problem in tb snapshot we are using now.

**Comment by Betterbird** [comment]
Sorry, I forgot to mention that FlatPak is has build 115 based on tag 115.0-bb6-build2. Perhaps they'll repeat based on115.0-bb6-build7.

**Comment by mfschumann** [comment]
I've built the flatpak successfully for BB 115.0-bb6-build6 using these build commands:
 https://github.com/flathub/eu.betterbird.Betterbird/blob/beta/eu.betterbird.Betterbird.json#L236

What rust version are you using for BB 115? I am using rust 1.70 for the flatpak.

**Comment by Betterbird** [comment]
FlatPak builds from a tarball of the source (500 MB) which is more efficient than pulling the repos (2 GB) every time. Building from the repos only makes sense on build machines which maintain the pulled repos and refresh them. The latter is the case for our builds, so we use https://github.com/Betterbird/thunderbird-patches/blob/main/build/build.sh. For one-off builds it's better to use something like https://github.com/Betterbird/thunderbird-patches/blob/main/build/build-one-off.sh. The latter didn't work due to https://bugzilla.mozilla.org/show_bug.cgi?id=1798746, but that's fixed now, so we'll need to see whether the script can be made to work. The problem was that `mach bootstrap` which sets up all the dependencies didn't work on the tarball.

FlatPak uses pre-configured builders, to the first approach makes more sense. Aren't you in the same boat?

**Comment by Betterbird** [comment]
We'll try building 115 on Linux with the build-one-off.sh in the next 24 hours. The script will need some fixing around line 200.

**Comment by mfschumann** [comment]
Does this approach work for builds between releases? I thought you were rebasing your patches onto the current TB source every now and then. Are there source tarballs for all intermediate commits in the TB repo?

**Comment by sagudev** [comment]
I was wondering what was build-one-off all about, but I didn't see a real difference when skimming them. But I am also wandering the same as [mfschumann](https://github.com/mfschumann), whether we always have it available.

**Comment by sagudev** [comment]
There is also problem with the scripts, when one command fails, execution continues, burying error in the logs.

**Comment by Betterbird** [comment]
> Does this approach work for builds between releases?

build-one-off.sh only works for fixed releases when TB make their release source available.

> I was wondering what was build-one-off all about, but I didn't see a real difference when skimming them.

The difference it pretty clear, one pulls the repos (2 GB), the other one the TB release tarball (500 MB)

> There is also problem with the scripts, when one command fails, execution continues, burying error in the logs.

Agreed. PR welcome.

We'll work on the build-one-off.sh script real soon now (TM) (but there's too much e-mail to answer, so we can't get to it).

**Comment by sagudev** [comment]
> build-one-off.sh only works for fixed releases when TB make their release source available.

Is BB always pinned to such release? If so than we can safely use build-one-off in CI else we cannot.

**Comment by Betterbird** [comment]
Always pinned, except for our "latest build" releases. But they are only for Windows, sometimes for Linux, but don't hit the "distros" (Arch, FlatPak). ... still answering e-mail, looks like it will need to wait until tonight.

**Comment by Betterbird** [comment]
OK, we fixed up the one-off script. `mach bootstrap` appeared to work fine. Sadly the script creates so much output that overruns the screen buffer, so it's hard to tell. Room for improvement.

https://github.com/Betterbird/thunderbird-patches/commit/df6c009296c93fe0cc6867e2271f54c3760ac2a4
https://github.com/Betterbird/thunderbird-patches/commit/d5e33160a6229a79e1917b64c14d5a1e841c36e8

Our build is running now.

**Comment by sagudev** [comment]
I have fails on patch https://github.com/Betterbird/thunderbird-patches/blob/main/115/features/feature-1802145-reestablish-movemail.patch as it is empty.

```console
 Applying patch feature-1802145-reestablish-movemail.patch ...
error: No valid patches in input (allow with "--allow-empty")
```

**Comment by Betterbird** [comment]
Yes, it's empty so far, not published yet. Strangely we got a

Applying patch feature-1802145-reestablish-movemail.patch ...
error: unrecognized input 

And it happily continued. For now, you'll need to deal with it (or we need to publish a dummy patch).

We'll look at the white-space errors in 12-feature-linux-systray.patch and 12-feature-linux-systray-example.patch. `git add` is just so terribly fussy. The stuff from 08-branding-m-c.patch we can't fix. That's a Windows script with CRLF and can't be changed.

**Comment by Betterbird** [comment]
Note this commit:
https://github.com/Betterbird/thunderbird-patches/commit/aed88ca77b3a8f0e6972cb14ff32369442a35a30

Our compile now fails on the libappinditator stuff. We'll have to see where that comes from. To be continued in six hours or so.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        haskell: true
+        large-packages: false
+        docker-images: true
+        swap-storage: false
+    - name: init env
+      run: |
+        source ${{ inputs.VERSION }}/${{ inputs.VERSION }}.sh
+        rustup override set $RUST_VER
+        sudo apt update
+        sudo apt install libdbusmenu-glib-dev libdbusmenu-gtk3-dev libdbusmenu-gtk-dev -y
```
Comment (sagudev) [inline .github/workflows/linux.yml:45]: Already counted on.

Comment (Betterbird) [reply .github/workflows/linux.yml:45]: Good. Maybe the extra ones fix my compile.

**Comment by mfschumann** [comment]
Please note #174 which should fix handling the empty movemail patch and makes the build script stop on errors.

**Comment by mfschumann** [comment]
> If anybody could test build artifact on ubuntu-20.04 please do and report if it is working.

I tested it in a Ubuntu 20.04 distrobox. Bettebird refused to start because it needs glibc 2.34, but Ubuntu 20.04 only provides glibc 2.31.

In a Ubuntu **22**.04 distrobox your Betterbird build started after installing libgtk-3-0, libasound2, libdbus-glib-1-2.

**Comment by sagudev** [comment]
In theory #174 should work, but in practice `-e` is hard to use right in bash scripts.

I have this commit that does this stuff on both scripts: https://github.com/sagudev/thunderbird-patches/commit/ac9f7407e7e4ea86e03aee460d43dba00e098bbe

Also do note that some commands are allowed to fail (`|| true`).

`-o pipefail` is actually not needed (currently) as there is no piping `|` in script.

**Comment by sagudev** [comment]
Now this PR have my latest work, but 115 are both failing. MacOS fails while catching rustc toolchain artifacts in bootstrap and Linux with error discussed above (that might be glibc problem).

**Comment by sagudev** [comment]
> I tested it in a Ubuntu 20.04 distrobox. Bettebird refused to start because it needs glibc 2.34, but Ubuntu 20.04 only provides glibc 2.31

I wonder how does tb and ff solves this issues as they are still shipping for ubuntu 20.04 right?

**Comment by sagudev** [comment]
After changing to 22.04, linux 115 alsp gets me appindicator errors:
https://github.com/sagudev/thunderbird-patches/actions/runs/5534808044/jobs/10100251063#step:6:1844

**Comment by Betterbird** [comment]
It's a real fight. Please look at our latest pushes. Everything applies cleanly now and that appindicator thing compiles.

**Comment by Betterbird** [comment]
Both FlatPak and us have now published a build based on 115.0-bb6-build8:
https://github.com/Betterbird/thunderbird-patches/releases/tag/115.0-bb6-build8

**Comment by Betterbird** [comment]
Let us know when you're ready. The links arriving via e-mail, like this one, don't work:
https://github.com/Betterbird/thunderbird-patches/pull/173/files/ac9f7407e7e4ea86e03aee460d43dba00e098bbe..06fafa83e927f90b057f38625623690a2cd93779

**Comment by sagudev** [comment]
> Let us know when you're ready. The links arriving via e-mail, like this one, don't work: https://github.com/Betterbird/thunderbird-patches/pull/173/files/ac9f7407e7e4ea86e03aee460d43dba00e098bbe..06fafa83e927f90b057f38625623690a2cd93779

That's just GitHub being terrible at handling force pushes.

**Comment by sagudev** [comment]
I think this is ready now.

Latest build: https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277/jobs/10117516404

**Comment by Betterbird** [comment]
Great. Will the be a manual on how to use it?

**Comment by sagudev** [comment]
Manual for what? CI pipeline runs on every commit and every pull request, thus you can test changes without building locally. You can [also dispatch workflows manually](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow).

- `Main` workflow runs all available workflows
- `Linux`/`Mac` workflow builds for ether 102 or 115, depending on the inputs (main workflow calls it twice; once for every build)

**Comment by Betterbird** [comment]
On every commit, please no. We can use this for releases, nothing else. 
Either when a label is added or when triggered manually.

Imagine fixing a typo in a commit and pulling 2 GB or 500 MB (tarball) 
from Mozilla for that to build.

**Comment by sagudev** [comment]
> On every commit, please no. We can use this for releases, nothing else.
Either when a label is added or when triggered manually.

Well this is also possible. Here is list of all events that could be set up to trigger CI: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows. We could then use `release` event.

> Imagine fixing a typo in a commit and pulling 2 GB or 500 MB (tarball) from Mozilla for that to build.

Well I really do not see this as a problem, as I would imagine Mozilla has these stuff in some sort of buckets, where they are only pay for storage, not for bandwidth. Actually [Mozilla's CI](https://treeherder.mozilla.org) is doing something similar for every commit push.

Also it's very common practice to have CI (or at least part of CI) running for every push/PR. Some project (ex. Rust) go even more stricter, they have [Bors](https://bors.tech/guide/2017/04/26/whirlwind/), which enforces that main branch is always buildable.

**Comment by Betterbird** [comment]
(back in the office)

> Well this is also possible. Here is list of all events that could be set up to trigger CI: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows. We could then use release event.

Yes, please. As I wrote, real useful for Mac M1/M2 builds. Are they already done?

I'm very well aware of Mozilla CI, I've worked in that space since 2015. I think it's wasteful, they build on a bazillion platforms after changing one comment :-( - They have DONTBUILD on the commit line which avoids building, but devs don't always remember.

For BB there are commits which will produce absolutely nothing new, for example:
https://github.com/Betterbird/thunderbird-patches/commit/75aa4ba5db682025d6ae632b6d763a73a706ccd1
https://github.com/Betterbird/thunderbird-patches/commit/ed085d94eed91535b068b961ca3d3e81e4ac676f
https://github.com/Betterbird/thunderbird-patches/commit/570ca357365fb80f42dd32b26a4744fa587d2985
etc.

So for this to be useful, it would need to be per tag or per manual trigger. 

Also, if I was told is right, you get 2000 build minutes free per month with Mac counting ten times as much (I haven't checked that information). So if we built on every commit, that would be exhausted soon.

More comments:

I understand that GitHub Actions provide a service to a project hosted on GitHub. Betterbird just hosts its Thunderbird modifications here, so it might not have been the intention of the creators of the service to pull giga tons of upstream code for 1% modification, even if the 1% make a lot of difference. So I'd use the service sparingly.

Looking at https://github.com/Betterbird/thunderbird-patches/compare/main...sagudev:thunderbird-patches:sagudev-patch-1, you're using the "one-off" build for 115 and the regular for 112 since the one-off doesn't work there, right? Hold, on, on Mac you're running the normal `build.sh` as a look into
https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277
confirms.

Why the indentation changes to `elif`?

Finally, where to download the Mac DMG? https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277/jobs/10117516404. Sorry if the question is silly, we're not CI people, we're C++, JS, XHTML and CSS people.

**Comment by sagudev** [comment]
> Also, if I was told is right, you get 2000 build minutes free per month with Mac counting ten times as much (I haven't checked that information). So if we built on every commit, that would be exhausted soon.

Well that it's true for private repos (if you have free GH account). But because we are on public repo we have unlimited minutes. The only catch is that you can have max 20 linux concurent jobs (5 if Mac) IIRC.

> GitHub Actions usage is free for standard GitHub-hosted runners in public repositories, and for self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage for use with GitHub-hosted runners, depending on the product used with the account. Any usage beyond the included amounts is controlled by spending limits. - https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#about-billing-for-github-actions

> I understand that GitHub Actions provide a service to a project hosted on GitHub. Betterbird just hosts its Thunderbird modifications here, so it might not have been the intention of the creators of the service to pull giga tons of upstream code for 1% modification, even if the 1% make a lot of difference. So I'd use the service sparingly.

I am not a layer, but as I understand this limit is mostly to prevent abuse like having code hosted somewhere else and have dummy repo on GH just for running free CI. Betterbird should have not problems here and also, there wouldn't be that much CI time spending as in some other projects.

**Comment by sagudev** [comment]
> Looking at [main...sagudev:thunderbird-patches:sagudev-patch-1](https://github.com/Betterbird/thunderbird-patches/compare/main...sagudev:thunderbird-patches:sagudev-patch-1), you're using the "one-off" build for 115 and the regular for 112 since the one-off doesn't work there, right? Hold, on, on Mac you're running the normal `build.sh` as a look into https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277 confirms.

Yes both 102 builds uses `build.sh`. Mac 115 also uses `build.sh` or else I get artifact errors: https://github.com/sagudev/thunderbird-patches/actions/runs/5534808044/jobs/10100251329#step:5:535

> Why the indentation changes to `elif`?

VsCode autoformatting, I should fix this.

> Finally, where to download the Mac DMG? https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277/jobs/10117516404. Sorry if the question is silly, we're not CI people, we're C++, JS, XHTML and CSS people.

Scroll down here: https://github.com/sagudev/thunderbird-patches/actions/runs/5542615277. There is mac-115 if you click it you get zip with dmg.

**Comment by Betterbird** [comment]
OK, can you please address the other questions:
1. Build on tag/manual only
2. Mac M1/2, yes, no, will do?
3. `elif` indentation changes
4. Where do download the artefacts (my GB spell-check doesn't like artifact).

**Comment by Betterbird** [comment]
> Mac 115 also uses build.sh or else I get artifact errors: https://github.com/sagudev/thunderbird-patches/actions/runs/5534808044/jobs/10100251329#step:5:535

Filed that as issue #175. Sad, so we're pulling those 2 GB all the time. 102.0 will be EOL soon.

**Comment by sagudev** [comment]
> 1. Build on tag/manual only

How about this:

```yml
on:
  # Run on tag
  release:
    types: [published]
  # Manual run
  workflow_dispatch:
  # Run on PR (per every push)
  pull_request:
```
> 2. Mac M1/2, yes, no, will do?

Currently it's a no. I do not have Mac, I fact the only knowledge I have on mac is from working on Servo's CI. But support for M1 in GitHub Actions is supposedly coming in 2023 Q3.

**Comment by Betterbird** [comment]
> How about this:

Sorry, I mentioned my skillset earlier on. I don't understand any of that syntax. Maybe it will run on tags. No idea how to trigger a manual push. And how go people get to the artefacts? I need to download them and publish them on the BB website? I did ask for a manual :wink:.

I guess there will be some notification about the builds.

Right I'd like to have a new build for Mac since there was a bug, but I can't sleep having to think about 2 GB being pulled for a 4-character-change. I guess I'll still build it locally for these occasions.

**Comment by sagudev** [comment]
Ignore push, but I am working on it.

**Comment by sagudev** [comment]
Okay so now CI runs on tags (depending on version contained in tag) and on manually dispatched.
Artifacts, as said before, are available on artifact section if you check the CI run. Ex. https://github.com/sagudev/thunderbird-patches/actions/runs/5550624292:
slika


I believe default retention of artifacts is 60 days. For tagged releases, we could actually uploaded them to as assets for GitHub Release. [Here](https://github.com/servo/servo-nightly-builds/releases/tag/2023-07-13) is how this looks on Servo project. These assets are stored forever and as far as I know there isn't any storage limit that we could hit.

**Comment by Betterbird** [comment]
OK. This looks good now. I still need a manual on how to trigger a build manually.

Building on tagging: OK, I'm still not sure where I want to do this in the early days of 115. I guess we can just comment-out this for now:
```
  push:
    tags: '[0-9][0-9][0-9].*'
```
Excuse the ignorance, but I can't see where the "depending on version contained in tag" is implemented.

Also, can we run Linux on Ubuntu 20 to get an older version of glibc for compatibility with users on older systems? We've declared Ubuntu 20 and Debian 11 the oldest supported systems [here](https://www.betterbird.eu/system-requirements/).

Thanks for all your work!

**Comment by sagudev** [comment]
Manual how to manually trigger workflow: https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow. But beware that until this PR lands there will be no buttons.

Depending on version is impl here: https://github.com/Betterbird/thunderbird-patches/pull/173/files#diff-21364b2e6fae1f2875cee1ab3daefb0685403687eaf8bc32b5c6eacda351c9d3R28-R31 and then every usage of `${{ env.VAR }}` uses this computed value.

As for building only on specific tags, it's easier to match this regex: '[0-9][0-9][0-9].*-build' for later parsing or something like that. So tell me how you would like it specifically and I will try to make it.

Unfortunately 115 does not build (due to glibc) on Ubuntu 22.04, so that would require to split workflow files.

**Comment by Betterbird** [comment]
> As for building only on specific tags, it's easier to match this regex: '[0-9][0-9][0-9].*-build' for later parsing or something like that. So tell me how you would like it specifically and I will try to make it.

Anything logical is fine. You could trigger a build when matching "^[0-9][0-9][0-9].*-automation". Then I push 115.0.1-bb7 or 115.0.1-bb7-build8-Mac-only, and I append "-automation" to trigger an automation build. I think manual requests won't be necessary after that.

> Unfortunately 115 does not build (due to glibc) on Ubuntu 22.04, so that would require to split workflow files.

Are you sure? I've built it locally on Mint 20 here, that used some older glibc that the Mozilla built system had downloaded and referenced in local `.mozbuild`. I had to wrestle with the compiler:
https://github.com/Betterbird/thunderbird-patches/commit/2dc6f36fb7ee4497a570f545c53f477b037347ca

Can I see the compilation errors. If we build it on Ubuntu 22, I can't distribute that since users on older systems will complain, as they have in the past.

**Comment by sagudev** [comment]
There are still problems https://github.com/sagudev/thunderbird-patches/actions/runs/5553263087/jobs/10141593486#step:7:1858 due to glibc.

**Comment by Betterbird** [comment]
Hmm, looks like it fails compiling a Rust thing, no?

```
Compiling cfg-if v0.1.999 (/home/runner/work/thunderbird-patches/mozilla-esr115/build/rust/cfg-if)
[snip]
error: failed to run custom build command for `glslopt v0.1.9`
```

BTW, you can see it's using its own sysroot `/home/runner/.mozbuild/sysroot-x86_64-linux-gnu/usr/include/x86_64-linux-gnu/bits/mathcalls.h`.

All I can say is that a build on Ubuntu 22 can't be distributed.

**Comment by sagudev** [comment]
This: https://searchfox.org/mozilla-central/source/third_party/rust/glslopt/glsl-optimizer/include/c99_math.h#188

**Comment by Betterbird** [comment]
I've taken this now. We can follow up later.

**Comment by sagudev** [comment]
`^` is not valid [Filter pattern](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet). What exactly should it do? Isn't `[0-9][0-9][0-9].*-automation` enough?

**Comment by Betterbird** [comment]
Thanks for the clarification, see:
https://github.com/Betterbird/thunderbird-patches/commit/1ab96ba93757bf37c20ed8bfb882b3768ddd486e

**Pull request merged**